inc/hclib-promise.h

```diff
@@ -59,7 +59,9 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 /*
  * The maximum number of future objects a task can wait on.
  */
+#ifndef MAX_NUM_WAITS
 #define MAX_NUM_WAITS 4
```

Comment:
Could you explain how you see this being used? Is this being added as a configure-time option?

Reply:
We need to build HClib with '-DMAX_NUM_WAITS=N' CXXFLAG.
For example to set the number to 6, build HClib as,
'HC_FLAGS_1=-DMAX_NUM_WAITS=6 ./install.sh'

Added as a part of the comment for this checkin.

Reply:
Can we just add this as a configure flag then? I didn't even realize that HC_FLAGS crud was still in there, it should be ripped out and we certainly shouldn't be adding more stuff that relies on it.

Reply:
Added a configure flag --with-max-dependency=[n]. It will show up with a brief explanation when using configure --help.

Reply:
Great! Can we delete the define now? For example, something like:

#ifndef MAX_NUM_WAITS
#error MAX_NUM_WAITS is not defined
#endif

Reply:
Why do we need to do that?
Currently we are using this default value set using #define if the user do not provide the  --with-max-dependency while building hclib. 
Also hclib.h includes hclib-promise.h, in which MAX_NUM_WAITS is defined, and therefore a user who uses hclib.h will get this error message if they dont define  MAX_NUM_WAITS while compiling their application.  
Otherwise we need to move code around so that those parts of the code not used by user should be invisible from the user.

Reply:
I agree that we need MAX_NUM_WAITS defined at all times. I didn't realize --with-max-dependency didn't provide a default value if it wasn't set. Is it possible to do that?

My concern is if HClib is built with one value of MAX_NUM_WAITS, and then an application is compiled with a different value #defined for MAX_NUM_WAITS, that conflict will be difficult to debug. Optimally, we want a single value set for MAX_NUM_WAITS at configure time which can't overridden accidentally at application compile time (or at least, if it is it requires someone actually going in and deleting code in header files, in which case we have to assume they know what they're doing).

Makes sense?

Reply:
That makes sense.  --with-max-dependency can provide default value but those values are available only during building HClib and will be used inside libhclib. Later when hclib.h is included in application program, it will  not get those macro values given during  building HClib (those values are only inside libhclib).
But I dont see any easy solution around that other than performing refactoring so that only values/structs etc required by the application programmer comes in exposed header files and others gets moved to some internal headers/source. For now can we change 

#ifndef MAX_NUM_WAITS
#define MAX_NUM_WAITS 4
#endif

to just 
#define MAX_NUM_WAITS 4

This way there is only point where MAX_NUM_WAITS is defined and for now if anyone wants to change it they need to edit the file and recompile, and create a bug/feature request to refactor the code?

Reply:
Wait, wouldn't that change just make --with-max-dependency unnecessary again?

Reply:
Yes. I wil remove it, other wise it will give compilation error that MAX_NUM_WAITS is redefined.

Reply:
Okay, so then the plan for any projects that need this > 4 (which I imagine yours might?) would be for them to edit this manually?

I'm fine with that, just want to make sure I'm understanding correctly.

The only big pain point on that front would be whenever we want to merge code in again. Any edits to MAX_NUM_WAITS might (or might not) have to be reverted first to keep it consistent on the master branch.

One future exploration might be to look at how this value affects microbenchmarks.

Reply:
Okay, sounds good. Can you make the change to delete these #ifndefs and the added configure stuff?

Once that's changed this looks good to me, but Zoran will have to sign off too (every merge requires 2 pairs of eyeballs). He also still needs to look at Akihiro's pull request.

@budimlic back from vacation yet?

Reply:
Great. Removed the ifndefs and configure flags.